Prevent counter values from over/underflowing

vitorroriz opened the pull request:
#### a3717f099f7779455b1f37625f208f6115dba449
<pre>
Prevent counter values from over/underflowing
<a href="https://bugs.webkit.org/show_bug.cgi?id=258730">https://bugs.webkit.org/show_bug.cgi?id=258730</a>
rdar://111743827

Reviewed by Chris Dumez.

We will ignore the counter-increment() operation if
this would result in overflow/underflow for compatibility
with other vendors (<a href="https://github.com/w3c/csswg-drafts/issues/9029).">https://github.com/w3c/csswg-drafts/issues/9029).</a>

 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow-expected.html: Added.
 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow-ref.html: Added.
 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow.html: Added.
 * Source/WTF/wtf/CheckedArithmetic.h:
 (WTF::sumIfNoOverflowOrFirstValue):
 * Source/WebCore/rendering/CounterNode.cpp:
 (WebCore::CounterNode::computeCountInParent const):

Canonical link: <a href="https://commits.webkit.org/266817@main">https://commits.webkit.org/266817@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b8064ffbdd11a9411f4a643c26587d5edd3186d9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14848 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/15152 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15505 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16596 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13974 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14985 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17673 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15253 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/16613 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/15028 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15514 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12609 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17329 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12793 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13394 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20371 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12706 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13872 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13562 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16801 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/14116 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/14113 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11926 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/15047 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13402 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/13251 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3840 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3581 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17735 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/15277 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13955 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3650 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/rendering/CounterNode.cpp:
```diff
@@ -139,11 +140,15 @@ CounterNode* CounterNode::previousInPreOrder() const
+    auto sumOrFirstValue = [](int firstValue, int secondValue) {
+        return WTF::signedAddInt32Overflows(firstValue, secondValue, secondValue) ? firstValue : secondValue;
+    };
+
```
@darinadler @cdumez Do we have some common pattern in the codebase for this? Blink seems to use a checkedAdd helper for instance.

darinadler replied:
We probably should not use `signedAddInt32Overflows`. That’s an implementation detail of `SaturatedArithmetic.h`, not designed for use outside the header. Ideally that header would be changed to just use `sumOverflows` from `CheckedArithmetic.h` to avoid duplication. But to preserve the performance `sumOverflows` would likely need to be specialized or optimized so it’s super efficient for two `int32_t` values. We’d want to do that, though. We don’t want two functions that do the same thing, one with a nicer name but a poorer implementation.

This all is in a saturated arithmetic header, and `saturatedSum` will clamp to the maximum `int32_t` value quite efficiently. But perhaps the counter specification requires we not saturate but just not add a new value at all if it would overflow. Too bad because saturated arithmetic is a nice solution for such things.

To improve the aesthetic of checked arithmetic it would be fine to evolve the checked arithmetic header to have the kind of function we need. We have `checkedSum` and `sumOverflows` and the `Checked` class template. Interestingly, `checkedSum` is _almost_ what we want, in fact it even stores the first value if adding the second value overflows. But you can’t get that value out without the code intentionally crashing in the `Checked::value` function!

I think that having `sumOrFirstValue` as a local lambda here is not great since this seems like the kind of operation we’d want to use more than once. I’d want it to be in `CheckedArithmetic.h`. We didn’t design that header, and it seems it’s falling short here. We may be able to come up with an even better name for that operation too. It’s possible that the `checkedAdd` function is something Chrome folks added to `CheckedArithmetic.h` after they forked WebKit for this very reason, but I don’t think that "add" vs. "sum" is very clear.

vitorroriz replied:
I've moved it to `CheckedArithmetic.h` and renamed it to `sumIfNoOverflow()`. I'm not sure about this name as it doesn't clearly tells what is returned if the sum won't take place. I'm open to suggestions!

cdumez commented on Source/WTF/wtf/CheckedArithmetic.h:
```diff
@@ -1013,6 +1013,12 @@ template<typename T, typename U> bool differenceOverflows(U left, U right)
     return checkedDifference<T>(left, right).hasOverflowed();
 }
 
+template<typename T> T sumIfNoOverflow(T left, T right)
```
I don't really like this function. It is really not obvious what's the right thing to return when the sum overflows. You currently decide to return `left` but it feels weird.

I'd rather call sites use the existing `checkedSum(left, right)` and then explicitly check for overflow.

vitorroriz replied:
As Darin mentioned:
`Interestingly, checkedSum is almost what we want, in fact it even stores the first value if adding the second value overflows. But you can’t get that value out without the code intentionally crashing in the Checked::value function!`

So, I could use `checkedSum()` locally and explicitly checking for overflow. I've done it similarly before using a lambda, but Darin asked it to be moved to a header. Maybe we need a better name though:

`I think that having sumOrFirstValue as a local lambda here is not great since this seems like the kind of operation we’d want to use more than once. I’d want it to be in CheckedArithmetic.h`

cdumez commented on Source/WebCore/rendering/CounterNode.cpp:
```diff
@@ -141,9 +142,9 @@ int CounterNode::computeCountInParent() const
     int increment = actsAsReset() ? 0 : m_value;
     if (m_previousSibling)
-        return m_previousSibling->m_countInParent + increment;
+        return WTF::sumIfNoOverflow(m_previousSibling->m_countInParent, increment);
```
I am a little surprised we ignore the increment in case of overflow (instead of clamping for e.g.) but since you said this was done to match other browsers, I'd recommend adding a comment in code explaining WHY we're doing this (ideally with the link to the spec issue too).

vitorroriz committed:
Prevent counter values from over/underflowing
https://bugs.webkit.org/show_bug.cgi?id=258730
rdar://111743827

Reviewed by Chris Dumez.

We will ignore the counter-increment() operation if
this would result in overflow/underflow for compatibility
with other vendors (https://github.com/w3c/csswg-drafts/issues/9029).

 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow-expected.html: Added.
 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow-ref.html: Added.
 * LayoutTests/imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow.html: Added.
 * Source/WTF/wtf/CheckedArithmetic.h:
 (WTF::sumIfNoOverflowOrFirstValue):
 * Source/WebCore/rendering/CounterNode.cpp:
 (WebCore::CounterNode::computeCountInParent const):

Canonical link: https://commits.webkit.org/266817@main
